### PR TITLE
[ELF] Orphan placement: prefer the last similar section when its rank <= orphan's rank

### LINKED ISSUE #92987: lld: Questionable heuristics for placing orphaned input sections
Hello,

The following was tested with LLVM/Clang version 17.0.6 and main branch.
This issue is similar, albeit not the same, as issue #92864 .

LLD seems to choose the first PHDR, in ascending order, with matching enough permissions to place orphaned input sections. This behaviour differs compared to ld.bfd, which prefers placing orphaned input sections with, e.g., `RW` permissions, into the PHDR which contains the `.data` output section, rather than the first PHDR that matches permissions in ascending order.

This is true whether the `.data` output section will be emitted or not, which also ties in with #92864 .

### PR DESCRIPTION
**Pull request by MaskRay**
`findOrphanPos` finds the most similar output section (that has input
sections). In the event of proximity ties, we select the first section.

However, when an orphan section's rank is equal to or larger than the
most similar sections's, it makes sense to prioritize the last similar
section. This new behavior matches GNU ld better.

```
// orphan placement for .bss (SHF_ALLOC|SHF_WRITE, SHT_NOBITS)

WA SHT_PROGBITS
(old behavior) <= here
A
WA SHT_PROGBITS
AX
WA (.data)
(new behavior) <= here
```

When the orphan section's rank is less, the current behavior
prioritizing the first section still makes sense.
```
// orphan with a smaller rank, e.g. .rodata

<= here
WA
AX
WA
```

Close #92987

### CODE REVIEW

**Review by smithp35 — COMMENTED** [review]
A few comments based on reading the code.

File: lld/ELF/Writer.cpp [hunk 1]
```diff
@@ -935,14 +935,18 @@ findOrphanPos(SmallVectorImpl<SectionCommand *>::iterator b,
...
     return e;
+  // Select the most similar output section. In case of ties, select the first
+  // section with a rank > the orphan's rank, or the last one with a rank <= the
+  // orphan's rank.
   int proximity = getRankProximity(sec, *b);
   auto i = b;
   for (auto j = b; ++j != e;) {
-    int p = getRankProximity(sec, *j);
-    if (p > proximity) {
+    auto p = getRankProximity(sec, *j);
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/Writer.cpp]: Done

File: lld/ELF/Writer.cpp [hunk 2]
```diff
@@ -935,14 +935,18 @@ findOrphanPos(SmallVectorImpl<SectionCommand *>::iterator b,
       return i;
   }
 
-  // Find the first element that has as close a rank as possible.
   if (b == e)
     return e;
+  // Select the most similar output section. In case of ties, select the first
+  // section with a rank > the orphan's rank, or the last one with a rank <= the
+  // orphan's rank.
   int proximity = getRankProximity(sec, *b);
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/Writer.cpp]: Thanks for the comment and variable name suggestion. Adopted.

The comment is slightly tuned.

File: lld/ELF/Writer.cpp [hunk 3]
```diff
@@ -935,14 +935,18 @@ findOrphanPos(SmallVectorImpl<SectionCommand *>::iterator b,
...
   auto i = b;
   for (auto j = b; ++j != e;) {
-    int p = getRankProximity(sec, *j);
-    if (p > proximity) {
+    auto p = getRankProximity(sec, *j);
+    if (p > proximity ||
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/Writer.cpp]: I used a variant of the comment that suggests the intention, but not replicate the "less" or "greater" behavior described by the code.
```
  // In the event of proximity ties, we select the first or last section
  // depending on whether the orphan's rank is smaller.
```

**Comment by MaskRay** [comment]
I have captured lld's algorithm in https://maskray.me/blog/2024-06-02-understanding-orphan-sections , which is hopefully useful.

**Review by igorkudrin — APPROVED** [review]
LGTM

**Review by smithp35 — APPROVED** [review]
Thanks for the update. Changes have made this easier to understand.

**Pull request merged**